src/formats/GRO.cpp

```diff
@@ -0,0 +1,324 @@
+/// not contain one more step.
+static bool forward(TextFile& file);
+
+GROFormat::GROFormat(const std::string& path, File::Mode mode)
+    : file_(TextFile::create(path, mode))
+{
+    while (!file_->eof()) {
+        auto position = file_->tellg();
+        if (!file_ || position == std::streampos(-1)) {
+            throw format_error("IO error while reading '{}' as GRO", path);
+        }
+        if (forward(*file_)) {
+            steps_positions_.push_back(position);
+        }
+    }
+    file_->rewind();
+}
+
+size_t GROFormat::nsteps() {
+    return steps_positions_.size();
+}
+
+void GROFormat::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+void GROFormat::read(Frame& frame) {
+    long count = 0;
+    try {
+        frame.set("name", (file_->readline())); // GRO comment line;
+        count = std::stol(file_->readline());
+    } catch (const std::exception& e) {
+        throw format_error("can not read next step as GRO: {}", e.what());
+    }
+
+    if (count < 0) {
+        throw format_error("the number of atoms can not be negative in GRO format");
+    }
+
+    size_t natoms = static_cast<size_t>(count);
+
+    residues_.clear();
+    frame.add_velocities();
+    frame.reserve(natoms);
+    frame.resize(0);
+
+    for (const auto& line: file_->readlines(natoms)) {
+        auto resid = std::stoul(line.substr(0,5));
+
+        std::string resname = line.substr(5, 5);
+
+        std::string atomname= line.substr(10,5);
+
+        // GRO files store atoms in NM, we need to convert to Angstroms
+        auto x = std::stof(line.substr(20,8)) * 10;
+        auto y = std::stof(line.substr(28,8)) * 10;
+        auto z = std::stof(line.substr(36,8)) * 10;
+
+        if (line.length() > 44) {
+            auto vx = std::stof(line.substr(44,8)) * 10;
+            auto vy = std::stof(line.substr(52,8)) * 10;
+            auto vz = std::stof(line.substr(60,8)) * 10;
+
+            frame.add_atom(Atom(trim(atomname)),
+                Vector3D(x, y, z),
+                Vector3D(vx, vy, vz)
+            );
+        } else {
+            frame.add_atom(Atom(trim(atomname)),
+                Vector3D(x, y, z)
+            );
+        }
+
+        auto cf_atomid = frame.size() - 1;
+        try {
+            if (residues_.find(resid) == residues_.end()) {
+                Residue residue(trim(resname), resid);
+                residue.add_atom(cf_atomid);
+
+                residues_.insert({resid, residue});
+            } else {
+                // Just add this atom to the residue
+                residues_.at(resid).add_atom(cf_atomid);
+            }
+        } catch (std::invalid_argument&) {
+            // No residue information
+        }
+
+    }
+
+    std::string box = file_->readline();
+    auto box_values = split(box, ' ');
+
+    if (box_values.size() == 3) {
+
+        auto a = std::stof(box_values[0]) * 10;
```

Comment:
Could you use `string2double` from utils.hpp instead of `std::stof`? The former check if the value is indeed a number and throw an error if not.

Reply:
Sure thing.